### PR TITLE
MAGECLOUD-4146: Cache warm-up script

### PR DESCRIPTION
**Pull request by dhorytskyi**
### Description
At the end of post-deploy phase the cache warm up script sends requests at a high rate such that weaker instances, like the 4-cpu ones, cannot cope. There nginx exhausts the number of workers, and this is without php-fpm, mysql or reds involvement.

This pull request adds the ability to specify the number of requests for warm-up which will be sent concurrently.

### Fixed Issues (if relevant)
https://magento2.atlassian.net/browse/MAGECLOUD-4146

### Manual testing scenarios
1. Specify a big list of pages to warm-up and the number of concurrent requests for warm-up.
In .magento.env.yaml add:
```
stage:
    post-deploy:
        WARM_UP_PAGES:
            - "index.php"
            ...
stage:
    post-deploy:
        WARM_UP_CONCURRENCY: 2
```
2. Deploy environment
3. Check the number of concurrent requests that are sending.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

releasenotes
Added the  [WARM_UP_CONCURRENCY post-deploy variable](https://devdocs.magento.com/cloud/env/variables-post-deploy.html) to specify the number of warm-up requests that the cache warm up script can send concurrently. Setting this option can help manage load on the Magento Cloud infrastructure.

### CODE REVIEW

File: src/Test/Unit/Step/PostDeploy/WarmUpTest.php [hunk 1]
```diff
@@ -84,9 +92,13 @@ public function testExecute()
             'http://base-url.com/index.php',
             'http://base-url.com/index.php/customer/account/create'
         ];
+        $concurrency = 0;
```
Comment (oshmyheliuk) [inline src/Test/Unit/Step/PostDeploy/WarmUpTest.php:95]: Please cover the case when concurrency is different than 0.

**Comment by YPyltiai** [comment]
I also wonder why we set concurrency to 0 by default. We need to find some middle ground between 0 and sending all at once, so it is faster but doesn't kill performance. @oshmyheliuk , do you think the code we used for SCD threads may work here too?

**Comment by dhorytskyi** [comment]
"0" is set by default to preserve current behavior if the new option is not set. The GuzzleHttp library, which is used under the hood, has own limit that is set to 25.
https://github.com/guzzle/guzzle/blob/master/src/Pool.php#L41

**Review by oshmyheliuk — APPROVED** [review]

**Review by shiftedreality — APPROVED** [review]

File: src/Test/Unit/Step/PostDeploy/WarmUpTest.php (hunk 1 above)
Comment (dhorytskyi) [reply src/Test/Unit/Step/PostDeploy/WarmUpTest.php:95]: Done

**Comment by andriyShevtsov** [comment]
@dhorytskyi @YPyltiai @oshmyheliuk maybe it make sense to add information about number of concurrency into `cloud.log`  otherwise I don't know how to check the number of concurrent requests. 
here is log with CONCURRENCY set to 3

```
[2020-03-18 23:03:42] DEBUG: Running step: warm-up
[2020-03-18 23:03:42] INFO: Starting page warming up
[2020-03-18 23:03:42] DEBUG: php ./bin/magento config:show:default-url --ansi --no-interaction
[2020-03-18 23:03:47] DEBUG: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/
[2020-03-18 23:03:51] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/privacy-policy-cookie-restriction-mode
[2020-03-18 23:03:51] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/sales/guest/form
[2020-03-18 23:03:51] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/contact
[2020-03-18 23:03:51] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/search/term/popular
[2020-03-18 23:03:52] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/index.php
[2020-03-18 23:03:52] DEBUG: Step "warm-up" finished
```

**Comment by andriyShevtsov** [comment]
QA is done. Waiting for response

**Comment by oshmyheliuk** [comment]
@andriyShevtsov I think you can test this PR without adding a log message. Just add more pages to warm-up and you should see that they are processing by 3 at a time. You have too small amount of pages to see this.
You can generate a small profile and warm-up all category or product pages, for example.

But adding a log message is a good Idea, something like 
``` Starting page warming up (concurrency: 3)```

**Review by bbatsche — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
@oshmyheliuk Are you referring to timestamps ?
```
[2020-03-19 18:20:14] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-85.html  
    [2020-03-19 18:20:14] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-84.html  
    [2020-03-19 18:20:14] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-86.html  
    [2020-03-19 18:20:15] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-87.html  
    [2020-03-19 18:20:15] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-88.html  
    [2020-03-19 18:20:15] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-89.html  
    [2020-03-19 18:20:16] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-91.html  
    [2020-03-19 18:20:16] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-90.html  
    [2020-03-19 18:20:16] INFO: Warmed up page: http://master-7rqtwti-zg62ybf3i2s3g.us-4.magentosite.cloud/simple-product-92.html  
```

 I suppose it should be something more obvious. But OK. I created separate task for this. 
https://magento2.atlassian.net/browse/MAGECLOUD-5636

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**